source/common/filesystem/kqueue/watcher_impl.cc

```diff
@@ -0,0 +1,170 @@
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/file_event.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+#include "common/filesystem/watcher_impl.h"
+
+#include "event2/event.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+WatcherImpl::WatcherImpl(Event::Dispatcher& dispatcher)
+    : queue_(kqueue()),
+      kqueue_event_(dispatcher.createFileEvent(queue_,
+                                               [this](uint32_t events) -> void {
+                                                 if (events & Event::FileReadyType::Read) {
+                                                   onKqueueEvent();
+                                                 }
+                                               },
+                                               Event::FileTriggerType::Edge,
+                                               Event::FileReadyType::Read)) {}
+
+WatcherImpl::~WatcherImpl() {
+  close(queue_);
+  watches_.clear();
+}
+
+void WatcherImpl::addWatch(const std::string& path, uint32_t events, Watcher::OnChangedCb cb) {
+  FileWatchPtr watch = addWatch_(path, events, cb, false);
+  if (watch == nullptr) {
+    throw EnvoyException(fmt::format("invalid watch path {}", path));
+  }
+}
+
+WatcherImpl::FileWatchPtr WatcherImpl::addWatch_(const std::string& path, uint32_t events,
+                                                 Watcher::OnChangedCb cb, bool path_must_exist) {
+  bool watching_dir = false;
+  int watch_fd = open(path.c_str(), O_SYMLINK);
+  if (watch_fd == -1) {
+    if (path_must_exist) {
+      return nullptr;
+    }
+
+    size_t last_slash = path.rfind('/');
+    if (last_slash == std::string::npos) {
+      throw EnvoyException(fmt::format("invalid watch path {}", path));
+    }
+
+    std::string directory = path.substr(0, last_slash);
+    watch_fd = open(directory.c_str(), 0);
+    if (watch_fd == -1) {
+      return nullptr;
+    }
+
+    watching_dir = true;
+  }
+
+  FileWatchPtr watch(new FileWatch());
+  watch->fd_ = watch_fd;
+  watch->file_ = path;
+  watch->events_ = events;
+  watch->callback_ = cb;
+  watch->watching_dir_ = watching_dir;
+
+  int flags = NOTE_DELETE | NOTE_RENAME;
+  if (watching_dir) {
+    flags = NOTE_DELETE | NOTE_WRITE;
+  }
+
+  struct kevent event;
+  EV_SET(&event, watch_fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, flags, 0,
+         reinterpret_cast<void*>(watch_fd));
+
+  if (kevent(queue_, &event, 1, NULL, 0, NULL) == -1) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(errno)));
+  }
+
+  if (event.flags & EV_ERROR) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(event.data)));
+  }
+
+  ENVOY_LOG(debug, "added watch for file: '{}' fd: {}", path, watch_fd);
+
+  watches_[watch_fd] = watch;
+
+  return watch;
+}
+
+void WatcherImpl::removeWatch(FileWatchPtr& watch) {
+  int fd = watch->fd_;
+  watches_.erase(fd);
```

Comment:
I was surprised that one didn't have to unregister but kevent docs note "Events which are attached to file descriptors are automatically deleted on the last close of the	descriptor."

Can you add a comment here that the deletion will close the fd and result in automatic unregistration?